src/passwordgeneration.js

```diff
@@ -16,33 +16,61 @@
         return password;
     };
 
-    upb.generate = function(location, difficulty, masterPassword, userSalt, callback, nolog) {
+    upb.generate = function(algorithm, location, difficulty, masterPassword, userSalt, callback, nolog, params) {
         if (!masterPassword) {
             throw new Error('master password should not be empty');
         }
 
-        var difficulty = difficulty || 8192;
-        if (!upb.isPowerOfTwo(difficulty)) {
-            throw new Error('difficulty should be a power of two, got ' + difficulty);
-        }
+        var host = typeof(location) === "string" ? location : location.protocol + '//' + location.host;
+        var userSalt = userSalt && userSalt != 0 ? "-keyidx:" + userSalt : ""; // keyidx is here for legacy reason, to avoid changing password
+        var salt = host + userSalt;
 
         var t = +new Date();
-        var logN = Math.log2(difficulty);
-        var host = location.protocol + '//' + location.host;
+        if(algorithm === 'scrypt') {
+            var difficulty = difficulty || 8192;
+            if (!upb.isPowerOfTwo(difficulty)) {
+                throw new Error('difficulty should be a power of two, got ' + difficulty);
+            }
+            var logN = Math.log2(difficulty);
 
-        var userSalt = userSalt && userSalt != 0 ? "-keyidx:" + userSalt : ""; // keyidx is here for legacy reason, to avoid changing password
-        var salt = host + userSalt;
+            scrypt(masterPassword, salt, logN, 8, 32, function(hashedPassword) {
+                var outputPassword = upb.makeHashHumanReadable(hashedPassword);
 
-        scrypt(masterPassword, salt, logN, 8, 32, function(hashedPassword) {
-            var outputPassword = upb.makeHashHumanReadable(hashedPassword);
+                if (!nolog && console && console.log) {
+                    var timeMessage = ' in ' + ((+new Date()) - t) / 1000 + ' seconds';
+                    var difficultyMessage = (difficulty > 1) ? ' (in ' + difficulty + ' difficulty)' : '';
+                    console.log('UniquePasswordBuilder - Generated password (scrypt): ' + outputPassword + ' for salt (domain + key index): ' + salt + timeMessage + difficultyMessage);
+                }
+                callback(outputPassword);
+            });
+        } else {
+            var difficulty = difficulty || 1;
+            // https://github.com/antelle/argon2-browser
+            argon2.hash({
+                pass: masterPassword,
+                salt: salt,
+                // optional
+                time: difficulty, // the number of iterations
+                // mem: 1024, // used memory, in KiB
+                hashLen: 32, // desired hash length
+                // parallelism: 1, // desired parallelism (will be computed in parallel only for PNaCl)
+                type: argon2.ArgonType.Argon2i, // or argon2.ArgonType.Argon2d
+                distPath: params === undefined ? './dist/' : params.argon2AsmPath
```

Comment:
@paulgreg 
There is here something painful with how argon2 works and how the code is built/deployed.

argon2 is composed of 2 files. The first one `argon2.js` loads the second one in asmjs `argon2-asm.min.js`. And, thus, in the configuration parameters, we must give the path toward this second file.

The problem is that between the 'addon' and the 'webpage', the js files are not at the same level and that's a pain to manage that (with the addon, we should look next to the `index.html` file. With the "webpage", we should look in the `dist` folder).

Is there a good reason for such a difference?
I am interested by flattening the "webpage" version. And, so, copy the `index.html` file in the `dist` folder.

That way, that could also make easier the deployment to github pages (that I try to do).

What do you think about it? 

The other solution is to pass the path through the `generate()` function parameters but that's a big pain due to an optional parameter `nolog` (perhaps I should switch the 2...)

Again, what is your point of view?